flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java

```diff
@@ -126,9 +140,14 @@ public synchronized void updateCurrentExecutionAttempts(Collection<JobDetails> j
                                                             taskMetricStore.getSubtaskMetricStore(
                                                                     subtaskIndex))
                                             .ifPresent(
-                                                    subtaskMetricStore ->
-                                                            subtaskMetricStore.retainAttempts(
-                                                                    attempts.getCurrentAttempts()));
+                                                    subtaskMetricStore -> {
+                                                        subtaskMetricStore.retainAttempts(
```

Comment:
Since the metrics for subtasks also exist in the taskMetricsStore in the form of taskInfo.subtaskIndex + "." + name, it is also necessary to clean up the transient metrics stored in the taskMetricsStore. Otherwise, it could lead to inconsistent behaviors that may confuse users, as depicted in the screenshot below.
![image](https://github.com/apache/flink/assets/107924572/ac8de662-7129-4445-a793-f0facca352e8)

Reply:
Good catch! But I'm wondering if this is a good approach to do so here. It complicates the code and makes maintenance more difficult.  Since the duplication was introduced to overcome the issue that WebInterface task metric queries currently do not account for subtasks, how about leaving it unremoved here for now and create a new JIRA on updating the WebInterface to account for subtasks? I'd be willing to help with that as well. cc @JunRuiLee @wanglijie95

Reply:
Thanks for clarification @X-czh .

I'm not quite certain I understand your concern.
In my opinion, this issue is unrelated to the web interface and more related to the inconsistency in the MetricStore because the WebUI is also get data from MetricStore. Specifically, the metrics in the subtaskMetricsStore are being removed, while the metrics in the taskMetricsStore are not synchronously removed, which could be confusing for users.

Based on your changes, you can perform the following test:
For a jobVertex that has already finished, you can use the JobVertexMetricsHandler to retrieve subtask metrics like below:
http://localhost:8081/jobs//vertices//metrics?get=0.backPressuredTimeMsPerSecond,0.busyTimeMsPerSecond
Then, compare the results with the SubtaskMetricsHandler:
http://localhost:8081/jobs//vertices//subtasks/0/metrics?get=backPressuredTimeMsPerSecond,busyTimeMsPerSecond

The results from these two endpoints are different. In my local test, the results are as shown in the attached image. I prefer that cleaning up should be done simultaneously for both, WDYT?

![image](https://github.com/apache/flink/assets/107924572/7ecdf3a0-f9ca-4ddf-aadd-a5d6a684f7f9)

Reply:
Hi @JunRuiLee, thanks for the clarification. I get your point and my concern is actual about maintainability of the code in the future. When there're both insertion and deletion operations, the duplication of task-level metrics actually makes it more difficult to maintain consistency (as is the case here), so I think it would be better to optimize the duplication issue here in the future.

I'll clean up simultaneously for both cases first here, and create a new issue for the optimization later. WDYT?

Reply:
Clean up simultaneously for both cases first here sounds good to me, Please feel free and go ahead~

Reply:
Updated, PTAL when you are free, thx~